ttbar and Hmumu analysis

IoPapadopoulos opened the pull request.

Soap2G reviewed (changes requested):
Hi @IoPapadopoulos, thanks for the update!

Here are a couple of suggestions coming from @alexander-held:

- the [vector](https://vector.readthedocs.io/) library would provide shortcuts for 4-vector math and simplify the code in some places, so might be worth a consideration
- the `uproot.open` call doesn't need the cycle (`key` in the notebook), it will just use the latest
- `cut_trig()` returns its argument so that function technically is not needed (but I understand if you want to keep the exact same interface for all cuts)
- `cut_charge()` can be simplified to `ak.sum(charge, axis=1) == 0`
- as far as I see, the dR cut for jets and muons will consider all jets, even those failing the jet selection cuts (same goes for the VBF cut), I'm not sure if this is on purpose

IoPapadopoulos commented:
Hi @Soap2G, thanks for the suggestions!

*I made the analysis run with parallel computing, reducing the runtime from 50 to 10 minutes.
*I tried to integrate the vector library into the code, but it keeps crashing. I’ll look into this again later.
*I fixed the issue with the key in uproot.open.
*I kept cut_trig() as is for consistency with the other cut functions.
*Thanks for the suggestion on cut_charge(), I updated it accordingly.
*Regarding the dR cut, yes, we need to check all the jets. For the VBF cut, I’ll double-check if it needs any adjustments.

Let me know if you have any other comments!

> Hi @IoPapadopoulos, thanks for the update!
> 
> Here are a couple of suggestions coming from @alexander-held:
> 
> * the [vector](https://vector.readthedocs.io/) library would provide shortcuts for 4-vector math and simplify the code in some places, so might be worth a consideration
> * the `uproot.open` call doesn't need the cycle (`key` in the notebook), it will just use the latest
> * `cut_trig()` returns its argument so that function technically is not needed (but I understand if you want to keep the exact same interface for all cuts)
> * `cut_charge()` can be simplified to `ak.sum(charge, axis=1) == 0`
> * as far as I see, the dR cut for jets and muons will consider all jets, even those failing the jet selection cuts (same goes for the VBF cut), I'm not sure if this is on purpose

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "# **The Goal: Hunt for the decay H $\\rightarrow \\mu^+ \\mu^-$**\n",
+        "This notebook will walk you through the process of possibly identifying and studying the decay of the Higgs boson into two muons ($ a\\ muon\\ and\\ an\\ anti-muon\\ \\mu^+ \\mu^-$). This decay channel is extremely rare, but it provides an essential probe of the interactions between the Higgs boson and second-generation fermions (leptons in this case) and muons in particular, since they  ...
+        "\n",
+        "By following this notebook, you will perform a $H \\rightarrow \\mu^+ \\mu^-$ search.You will apply a sequence of event selection criteria (cuts) to maximize the signal (events where $H \\rightarrow \\mu^+ \\mu^-$) to background (other processes producing pairs of muons) ratio.\n",
```
```suggestion
        "By following this notebook, you will perform a $H \\rightarrow \\mu^+ \\mu^-$ search. You will apply a sequence of event selection criteria (cuts) to maximize the signal (events where $H \\rightarrow \\mu^+ \\mu^-$) to background (other processes producing pairs of muons) ratio.\n",
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "import matplotlib.pyplot as plt # For creating plots and visualizations\n",
+        "from matplotlib.ticker import AutoMinorLocator # for minor ticks\n",
+        "from concurrent.futures import ProcessPoolExecutor, as_completed # Enables parallel execution for faster processing of large datasets\n",
+        "import warnings\n",
```
```suggestion
        "# Filter warnings that otherwise appear in output. These are normal in the running of this notebook.\n",
        "import warnings\n",
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+      "outputs": [],
+      "source": [
+        "# Integrated luminosity in inverse femtobarns\n",
+        "lumi = 10000.\n",
```
Should this now be 36.? Also, above we talk about definitions in inverse femtobarns, but here I think we're in inverse picobarns? Maybe a comment on the units of this number?

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+      },
+      "source": [
+        "# **Samples**\n",
+        "We select these samples below to capture both the signal and relevant background processes for the Higgs boson decay to muon pairs ($H \\rightarrow \\mu^+ \\mu^-$). The data samples include all possible processes that produce a pair of muon-antinuon.  The background samples consist of simulated events chosen to account for the primary sources of noise in the analysis: the continuum Drell ...
```
```suggestion
        "We select these samples below to capture both the signal and relevant background processes for the Higgs boson decay to muon pairs ($H \\rightarrow \\mu^+ \\mu^-$). The data samples include all possible processes that produce a pair of muon-antinuon.  The background samples consist of simulated events chosen to account for the primary sources of noise in the analysis: the continuum Drell-Yan $\\mu^+ \\mu^-$production, which closely mimics the signal, and smaller contributions from $t\\bar{t}$ production and diboson or $W + jets$ processes. Including these samples allows us to model the background accurately and optimize cuts to enhance the signal-to-background ratio.",
        "Notice that we are using a 2 muon skim. Events have been pre-selected to include two good muons to reduce the total amount of data that must be processed."
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "    * Identification (ID) Check:\n",
+        "  The function checks whether the total number of particles that pass the identification criteria in each event is greater than one. This ensures that the event contains at least two well-identified muons.\n",
+        "    * Isolation (Iso) Check:\n",
+        "  Similarly, the function checks whether the number of particles passing the isolation criteria is greater than one for each event. This ensures that multiple particles in the event are well-isolated.Non-isolated muons could be non-prompt ones coming from adronic decays."
```
```suggestion
        "  Similarly, the function checks whether the number of particles passing the isolation criteria is greater than one for each event. This ensures that multiple particles in the event are well-isolated. Non-isolated muons could be non-prompt ones coming from hadronic decays."
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "    1. **Energy Check:** Confirms that the event has at least two jets with an energy of 30 GeV or higher.\n",
+        "    2. **Transverse Momentum Check:** Verifies that at least two jets have transverse momentum values exceeding 30 GeV or higher.\n",
+        "    3.**b-Tagging Check:**\n",
+        "      Ensures that at least two jets have b-tagging scores less than or equal to 3. Lower b-tagging scores indicate a higher probability of the jet contains b-quarks from decays of top quarks,which could contain non-prompt background muons.\n",
```
```suggestion
        "      Ensures that at least two jets have b-tagging scores less than or equal to 3. Lower b-tagging scores indicate a higher probability of the jet contains b-quarks from decays of top quarks, which could contain non-prompt background muons (e.g. muons from hadron decays, rather than from Higgs bosons).\n",
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+      "outputs": [],
+      "source": [
+        "def cut_DR(eta,phi,eta_M,phi_M):\n",
+        "    DR_M1 = np.sqrt((eta - eta_M[:,0])**2 + (phi -  phi_M[:,0])**2)\n",
```
This is fine for delta eta, but phi is periodic. What happens when a jet has phi=3.1 and a muon has phi=-3.1, and what should happen?

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "  * The cut also assesses the separation in pseudorapidity between jet pairs. Specifically, it checks if the absolute difference in $\\eta$ between at least one pair of jets is greater than 3, indicating they are well-separated in rapidity.\n",
+        "  * Additionally, it ensures that at least one pair of jets has one jet with a positive $\\eta$ and the other with a negative $\\eta$, confirming the jets are produced in opposite directions.\n",
+        "\n",
+        "* Alternate Selection Option: Consider relaxing the VBF cuts (e.g. lowering the invariant mass) or switching to a ggF-based selection."
```
```suggestion
        "* Alternate Selection Option: Consider relaxing the VBF cuts (e.g. lowering the invariant mass) or switching to a ggF-based selection. ggF normally does not have significant jet activity, so one can try a jet veto instead."
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+      },
+      "source": [
+        "8. **Jet Selection Cut:**\n",
+        "  * This selection cut ensures that the event contains at least two jets that satisfy specific criteria for energy, transverse momentum, and b-tagging. The two jets come from the production of the Higgs boson, either by the Gluon Fusion or the Vector Boson Fusion mechanisms, which are the dominant production ones.\n",
```
```suggestion
        "  * This selection cut ensures that the event contains at least two jets that satisfy specific criteria for energy, transverse momentum, and b-tagging. The two jets come from the production of the Higgs boson, normally in the Vector Boson Fusion mechanism, which is one of the dominant production mechanisms. The gluon-gluon fusion (ggF) production mechanism is more common, but can be harder to distinguish from the background because of its relatively lower jet activity.\n",
```

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,947 @@
+        "            prefix = \"Data/\"\n",
+        "        else: prefix = \"MC/\"\n",
+        "        #################### change to the final link #######################\n",
+        "        tuple_path = \"https://cernbox.cern.ch/files/spaces/eos/user/e/egramsta/OpenData/NOV2024/2muons/\"\n",
```
This is what we should update to use the 'final' location of the files.

zlmarshall reviewed (commented):
Thanks very much for this MR! I've tried to add a few suggestions that I hope will help the clarity of the documentation. You did a very nice job, I think!

Cheers,
Zach

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,997 @@
+        "    # Add text to the plot\n",
+        "    main_axes.text(0.05, 0.93, 'ATLAS Open Data', transform=main_axes.transAxes, fontsize=13)\n",
+        "    main_axes.text(0.05, 0.88, 'for education', transform=main_axes.transAxes, style='italic', fontsize=8)\n",
+        "    #main_axes.text(0.05, 0.82, r'$\\sqrt{s}$=13 TeV, 10 fb$^{-1}$', transform=main_axes.transAxes)\n",
```
Still 10/fb?

IoPapadopoulos replied:
No, it should be. thirty 36 fb.

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,997 @@
+        "        if data_type == \"data\":\n",
+        "            prefix = \"Data/\"\n",
+        "        else: prefix = \"MC/\"\n",
+        "        #################### change to the final location of the files #######################\n",
```
Reminder that we can use `root://eospublic.cern.ch//eos/opendata/atlas/rucio/user/egramsta/` as the path to the files now

IoPapadopoulos replied:
I did update the path to the files.

zlmarshall commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,997 @@
+      "source": [
+        "def cut_DR(eta,phi,eta_M,phi_M):\n",
+        "\n",
+        "    Dphi1 = phi -  phi_M[:,0]\n",
```
I think this still has the same phi-periodicity problem? e.g. if `phi=-3.1` and `phi_M=3.1`, then `Dphi1=-6.2`, even though they are only 0.08 units apart? One common trick is to use `asin` and `sin`, like: `dPhi = asin( sin( phi-phiM ) )`, which has the right domain.

IoPapadopoulos replied:
I went ahead and updated the equation to use the expression you suggested

zlmarshall commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -83,25 +70,25 @@
-      "outputs": []
+      "outputs": [],
+      "source": [
+        "!pip install uproot"
```
Standard question for @Soap2G as to whether this is needed

IoPapadopoulos replied:
The google colab  does not come with the uproot to installed. So I have to install it every time. This is why it's the first thing on the notebook.

zlmarshall reviewed (commented):
Thanks for the updates! Just a few questions about the new files...

Best,
Zach

IoPapadopoulos commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -619,347 +597,334 @@
+        "            prefix = \"Data/\"\n",
+        "        else: prefix = \"MC/\"\n",
+        "        #################### change to the final link #######################\n",
+        "        tuple_path = \"https://cernbox.cern.ch/files/spaces/eos/user/e/egramsta/OpenData/FEB2025/1LMET30/\"\n",
```
I did update the path to the files.

zlmarshall commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -83,25 +70,27 @@
+      "outputs": [],
+      "source": [
+        "!pip install uproot\n",
+        "!pip install fsspec-xrootd\n",
```
@marianaiv and @Soap2G we should agree on how to handle this I think

Soap2G replied:
Also, https://github.com/atlas-outreach-data-tools/atlasopenmagic/pull/10

zlmarshall replied:
Just so I'm sure, that PR (with atlasopenmagic 0.4.0 when it's released) would change the code into something more like:

```
import subprocess,sys
subprocess.run([sys.executable, "-m", "pip", "install", "--upgrade", "--user","atlasopenmagic"])
from atlasopenmagic import install_from_environment
required_packages = ['coffea', 'hist', 'dask', 'numpy', 'matplotlib']
install_from_environment(required_packages, environment_file="../../binder/environment.yml")
```

?

Soap2G replied:
Yeah, it could also be 
```python
!pip install atlasopenmagic
from atlasopenmagic import install_from_environment
install_from_environment('coffea', 'hist', 'dask', 'numpy', 'matplotlib', environment_file="../../binder/environment.yml")
```

There is still some convolution, but at least one doesn't have to worry about versions, and everything depends from environment.yml

zlmarshall replied:
Just a note to say that this should now be working with the other merges done today. That should make your life a bit simpler :)

marianaiv replied:
This should now be:
```
from atlasopenmagic import install_from_environment
install_from_enviroment()
```

zlmarshall commented:
Thanks @IoPapadopoulos ! It looks good to me at this point. We just have to make sure we're consistent on how to handle the file reading from the open data portal (was just talking to Mariana and Giovanni about that today, so I think we have a solution, or 90% of one)

marianaiv commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,994 @@
+      },
+      "outputs": [],
+      "source": [
+        "import uproot # For reading ROOT files efficiently\n",
+        "import awkward as ak # To represent nested data in columnar format\n",
+        "import pandas as pd # For dataframes, a format widely used in python\n",
+        "import numpy as np # For numerical calculations such as histogramming\n",
+        "import time # For timing operations and adding delays if needed\n",
+        "import matplotlib.pyplot as plt # For creating plots and visualizations\n",
+        "from matplotlib.ticker import AutoMinorLocator # for minor ticks\n",
+        "from concurrent.futures import ProcessPoolExecutor, as_completed # Enables parallel execution for faster processing of large datasets\n",
+        "# Filter warnings that otherwise appear in output. These are normal in the running of this notebook.\n",
+        "import warnings\n",
```
Now we have a function in atlasopenmagic to setup the notebooks
```
import atlasopenmagic as atom
atom.install_from_environment()
```
This should be enough, it would install all in the environment file from this repository. If you want to add a package that is not in the environment, you can do it listing them:
```
atom.install_from_environment('package1', 'package2')
```
Nothing should be missing from the environment though, so let us know is that is the case

marianaiv commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,994 @@
+      },
+      "outputs": [],
+      "source": [
+        "samples = {\n",
+        "\n",
+        "    'data': {'list' : [\"data15_periodD\",\"data15_periodE\",\"data15_periodF\",\"data15_periodG\",\n",
+        "                        \"data15_periodH\",\"data15_periodJ\",\"data16_periodA\",\"data16_periodB\",\n",
+        "                        \"data16_periodC\",\"data16_periodD\",\"data16_periodE\",\"data16_periodF\",\n",
+        "                        \"data16_periodG\",\"data16_periodI\",\"data16_periodK\",\"data16_periodL\"],},\n",
+        "\n",
+        "    'mumu' : {  'list' : [\"mc_700323.Sh_2211_Zmumu_maxHTpTV2_BFilter\",\n",
+        "                          \"mc_700324.Sh_2211_Zmumu_maxHTpTV2_CFilterBVeto\",\n",
+        "                          \"mc_700325.Sh_2211_Zmumu_maxHTpTV2_CVetoBVeto\"\n",
+        "                        ],},\n",
+        "\n",
+        "    'ttbar' : { 'list' : ['mc_410470.PhPy8EG_A14_ttbar_hdamp258p75_nonallhad'],},\n",
+        "\n",
+        "    'Higgs' : { 'list' : ['mc_345097.PowhegPythia8EvtGen_NNLOPS_nnlo_30_ggH125_mumu',\n",
+        "                          'mc_345098.PowhegPythia8EvtGen_NNPDF3_AZNLO_ggZH125_Hmumu_Zinc',\n",
+        "                          'mc_345106.PowhegPythia8EvtGen_NNPDF30_AZNLOCTEQ6L1_VBFH125_mumu' ],}\n",
+        "                        }"
```
For the samples we also now have a helper function. You can use it like:
```
mc_defs = {
    r'Background $t\bar t$':    {'dids': [410470],                      'color': 'yellow'},
    r'Background $V+$jets':     {'dids': [700335,700336,700337],        'color': 'orange'},
    r'Background Diboson':      {'dids': [700488,700489,700490,700491],'color': 'green'},
    r'Background $ZZ^{*}$':     {'dids': [700600,700601],               'color': '#ff0000'},
    r'Signal ($m_H$=125 GeV)':  {'dids': [345060,346228],              'color': '#00cdff'},
}

mc_samples   = build_mc_dataset(mc_defs,   skim='2bjets', protocol='https')
data_samples = build_data_dataset('2bjets',  color='red', protocol='https')

samples = {**data_samples, **mc_samples}
```
the color is optional.

This should give you the dict with all the urls you will need to stream the data.

marianaiv commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,994 @@
+      "source": [
+        "  3. **Dilepton Selection:**\n",
+        "\n",
+        "    *   This cut ensures that the event contains exactly two leptons, which is essential for isolating signal events involving a muon pair.\n",
+        "\n",
+        "    *   Purpose: The cut checks if the number of leptons in the event (lep_n) is exactly equal to 2. Events with two leptons are selected for further analysis, as this is a signature of the $H \\rightarrow \\mu^+ \\mu^-$ decay, where the Higgs boson decays into a pair of muons. Events with fewer or more leptons are rejected to reduce background events."
```
For some reason this is not rendering correctly. Maybe is indented more than once? that's markdown signal for code

IoPapadopoulos replied:
Fixed it.

marianaiv commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,994 @@
+        "id": "cpLMrzk_VB5I"
+      },
+      "execution_count": null,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "kfzn5IsUgD-Q"
+      },
+      "source": [
+        "# **Data Aggregation with Loops**\n",
+        "The code below runs over multiple datasets and concatenates the results into Pandas DataFrames. It performs this for a specified number of iterations (loops) to aggregate data from multiple files.\n",
+        "\n",
+        "This loop-based approach allows the processing of data in batches, making it scalable and memory-efficient for handling large datasets."
+      ]
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "9Nubi7VggE4b"
+      },
+      "outputs": [],
+      "source": [
+        "Number_of_loops = 20 # If you want the analysis to go faster reduce this variable (always with an integer)\n",
+        "\n",
+        "print(\"The parallel analysis has started \\n\")\n",
+        "start_time = time.time()\n",
+        "\n",
+        "i = 0\n",
+        "\n",
+        "while i < Number_of_loops:\n",
+        "    Data_df = pd.concat([Data_df, parallel_analysis(\"data\",i)], ignore_index=True)\n",
+        "    mumu_df = pd.concat([mumu_df, parallel_analysis(\"mumu\",i)], ignore_index=True)\n",
+        "    ttbar_df = pd.concat([ttbar_df, parallel_analysis(\"ttbar\",i)], ignore_index=True)\n",
+        "    Higgs_df = pd.concat([Higgs_df, parallel_analysis(\"Higgs\",i)], ignore_index=True)\n",
+        "    i += 4\n",
+        "    elapsed_time = time.time() - start_time\n",
+        "    print(f\"{round(i*fraction*100)}% of the analysis completed in {round(elapsed_time / 60, 1)} min \\n\")"
```
Creating empty dataframes and filling in a loop is very memory inefficient. I suggest to fill dictionaries and concatenate at the end. Something like:
```
samples = ["data", "mumu", "Higgs", "ttbar"]
results = {s: [] for s in samples}

# Loop over chunks and append to lists
for i in range(number_of_loops):
    for sample in samples:
        chunk_df = parallel_analysis(sample, i)
        results[sample].append(chunk_df)
    elapsed = time.time() - start_time
    percent = (i + 1) / number_of_loops * 100
    print(f"{percent:.0f}% done in {elapsed/60:.1f} min")

# Single concat 
Data_df  = pd.concat(results["data"],  ignore_index=True)
mumu_df  = pd.concat(results["mumu"],  ignore_index=True)
Higgs_df = pd.concat(results["Higgs"], ignore_index=True)
ttbar_df = pd.concat(results["ttbar"], ignore_index=True)
```

marianaiv commented on 13-TeV-examples/uproot_python/Hmumu.ipynb:
```diff
@@ -0,0 +1,994 @@
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
```
Would be good to have the notebook with the outputs to check what we are getting here. You just need to save it after running it and commit/push so that it is uploaded with all the outputs.

marianaiv commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -166,142 +152,122 @@
+      "outputs": [],
+      "source": [
+        "# Dictionary of samples to be processed\n",
+        "samples = {\n",
```
Same comment here as for the Hmumu notebook.
```
import atlasopenmagic as atom
mc_defs = {
    r'Background $t\bar t$':    {'dids': [410470],                      'color': 'yellow'},
    r'Background $V+$jets':     {'dids': [700335,700336,700337],        'color': 'orange'},
    r'Background Diboson':      {'dids': [700488,700489,700490,700491],'color': 'green'},
    r'Background $ZZ^{*}$':     {'dids': [700600,700601],               'color': '#ff0000'},
    r'Signal ($m_H$=125 GeV)':  {'dids': [345060,346228],              'color': '#00cdff'},
}

mc_samples   = atom.build_mc_dataset(mc_defs,   skim='2bjets', protocol='https')
data_samples = atom.build_data_dataset('2bjets',  color='red', protocol='https')

samples = {**data_samples, **mc_samples}
```
Something like this.

marianaiv commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -312,162 +278,180 @@
+        "\n",
+        "   *    This cut ensures that the event contains one lepton that is matched with a trigger.\n",
+        "\n",
+        "   *     Purpose: The cut ensures that one lepton in the event is associated with a trigger, meaning it was responsible for the event being recorded. Applying this cut reduces backgrounds from events where no lepton is properly matched to a trigger, improving the selection efficiency for signal events."
```
This bullet point also has a weird font. Maybe it has extra spaces and it is taken as code?

marianaiv commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -619,347 +600,334 @@
-        "# Combine mtop2 and mtop3 lists main_analysis_df\n",
-        "mtop2 = ak.flatten([main_analysis_df[\"mtop2\"], main_analysis_df[\"mtop3\"]])\n",
-        "weights_data = ak.Array([1/2] * 2 * len(main_analysis_df))\n",
+        "Data_df = pd.DataFrame()\n",
```
Same comment here as in the other notebook. Creating empty dataframes and filling in a loop is slow.

marianaiv commented on 13-TeV-examples/uproot_python/ttbar_analysis.ipynb:
```diff
@@ -1021,55 +983,55 @@
     },
     {
       "cell_type": "code",
+      "execution_count": null,
```
And same comment, please upload it with the outputs :)

marianaiv reviewed (changes requested):
Hi @IoPapadopoulos!

Thank you for the work. I really appreciate the documentation on these. Added some comments, mostly about using atlasopenmagic, and one for the looping and saving cuts.

zlmarshall commented:
Hi @IoPapadopoulos ,

What's the problem with the output in the analyses?

Best,
Zach

IoPapadopoulos commented:
Hi @marianaiv,

I have addeed the following lines in the start of both analyses.

```
#install required packages
import sys
!pip install atlasopenmagic
from atlasopenmagic import install_from_environment
install_from_environment()
```
If you have any further comments or suggestions, please feel free to let me know.

Best, 
Giannis

marianaiv reviewed (approved):
Thanks @IoPapadopoulos, this is great!
For me this is good to go.
